util-core/src/main/scala/com/twitter/util/Batcher.scala

```diff
@@ -0,0 +1,28 @@
+package com.twitter.util
+
+import java.util.logging.{Level, Logger}
+
+import scala.collection.mutable
+
+class Batcher[In, Out](
+  sizeThreshold: Int,
+  timeThreshold: Duration = Duration.Top,
+  sizePercentile: => Float = 1.0f,
+  f: Seq[In] => Future[Seq[Out]]
+)(
+  implicit timer: Timer
+) extends Function1[In, Future[Out]] { batcher =>
+  import java.util.logging.Level.WARNING
+
+  val executor = new BatchExecutor[In, Out](sizeThreshold, timeThreshold, sizePercentile, f)
+
+  def apply(t: In): Future[Out] = executor.enqueue(t)
+
+  def flushBatch() = {
+    val doAfter = synchronized {
```

Comment:
Testing for concurrency is very difficult, it's easier to just reason about the java memory model.  If we start running into big problems, we can go write some multithreaded tests to try to trigger the race condition, but it's fine for now to stick with singlethreaded tests.

With that said, the guarantee that we want is that there is only one caller of flushBatch() at a time.  The way `BatchExecutor` achieves this is by synchronizing on itself, which creates a happens-before relationship between any two calls to flushBatch().  I think the only way for us to do the same thing is to synchronize on the same thing (ie the `BatchExecutor`).